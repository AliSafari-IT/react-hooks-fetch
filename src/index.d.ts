type Input<Data> = string | Request | (() => Promise<Data>);

export type CreateResource = <Data>(input: Input<Data>) => {
  data: Data;
};

export type CreateEmptyData = <Data>(emptyData: Data) => {
  data: Data;
};

export type UseResource = <Data>(
  initialResource: { data: Data }
) => {
  data: Data;
  refetch: (input: Input<Data>) => void;
};

export const createResource: CreateResource;
export const createEmptyData: CreateEmptyData;
export const useResource: UseResource;
