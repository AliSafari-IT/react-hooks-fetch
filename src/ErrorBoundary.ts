import { Component, ReactNode } from 'react';

type Fallback = ReactNode | ((error?: Error, retry?: () => void) => ReactNode);

type Props = {
  fallback: Fallback;
};

type State = {
  error: Error | null;
};

/**
 * ErrorBoundary with retry support
 *
 * @example
 * import { ErrorBoundary } from 'react-hooks-fetch';
 *
 * const App = () => (
 *   <ErrorBoundary
 *     fallback={({ error, retry }) => (
 *       <div>
 *         <span>{error.message}</span>
 *         <button type="button" onClick={retry}>Retry</button>
 *       </div>
 *     )}
 *   >
 *     ...
 *   </ErrorBoundary>
 * };
 */
export class ErrorBoundary extends Component<Props, State> {
  state = { error: null } as State;

  static getDerivedStateFromError(error: Error) {
    return { error };
  }

  retry = () => { this.setState({ error: null }); };

  render() {
    const { error } = this.state;
    const { children, fallback } = this.props;
    if (error) {
      if (typeof fallback === 'function') {
        return fallback(error, this.retry);
      }
      return fallback;
    }
    return children;
  }
}
