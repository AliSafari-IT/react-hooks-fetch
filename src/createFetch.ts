type FetchFunc<Input, Result> = (input: Input) => Promise<Result>;

export type FetchStore<Input, Result> = {
  prefetch: (input: Input) => void;
  evict: (input: Input) => void;
  use: (input: Input, mark?: symbol) => () => void;
  getResult: (input: Input) => Result;
};

type UsedMarks = Set<symbol>;

type Options<Input, Result> = {
  preloaded?: Iterable<{ input: Input; result: Result }>;
  areEqual?: (a: Input, b: Input) => boolean;
};

/**
 * create fetch store
 *
 * @example
 * import { createFetch } from 'react-hooks-fetch';
 *
 * const fetchFunc = async (userId) => {
 *   const res = await fetch(`https://reqres.in/api/users/${userId}?delay=3`));
 *   const data = await res.json();
 *   return data;
 * };
 * const store = createFetch(fetchFunc);
 * store.prefetch('1');
 */
export function createFetch<Input, Result>(
  fetchFunc: FetchFunc<Input, Result>,
  options?: Options<Input, Result>,
) {
  const preloaded = options?.preloaded;
  const areEqual = options?.areEqual;
  type GetResult = () => Result;
  const cache = new Map<Input, [GetResult, UsedMarks]>();
  const setCache = (input: Input, getResult: GetResult) => {
    if (areEqual) {
      for (const key of cache.keys()) {
        if (areEqual(key, input)) {
          return false;
        }
      }
    } else if (cache.has(input)) {
      return false;
    }
    cache.set(input, [getResult, new Set()]);
    return true;
  };
  const getCache = (input: Input) => {
    if (areEqual) {
      for (const [key, value] of cache) {
        if (areEqual(key, input)) {
          return value;
        }
      }
      return undefined;
    }
    return cache.get(input);
  };
  const deleteCache = (input: Input) => {
    if (areEqual) {
      for (const key of cache.keys()) {
        if (areEqual(key, input)) {
          cache.delete(key);
          return true;
        }
      }
      return false;
    }
    return cache.delete(input);
  };
  if (preloaded) {
    for (const item of preloaded) {
      setCache(item.input, () => item.result);
    }
  }
  const createGetResult = (input: Input) => {
    let promise: Promise<void> | null = null;
    let result: Result | null = null;
    let error: unknown | null = null;
    promise = (async () => {
      try {
        result = await fetchFunc(input);
      } catch (e) {
        error = e;
      } finally {
        promise = null;
      }
    })();
    const getResult = () => {
      if (promise) throw promise;
      if (error !== null) throw error;
      return result as Result;
    };
    return getResult;
  };
  const prefetch = (input: Input) => {
    setCache(input, createGetResult(input));
  };
  const evict = (input: Input) => {
    deleteCache(input);
  };
  const use = (input: Input, mark = Symbol()) => {
    const usedMarks = getCache(input)?.[1];
    if (!usedMarks) throw new Error('call prefetch in advance');
    usedMarks.add(mark);
    return () => {
      // check if it's already evicted
      if (usedMarks === getCache(input)?.[1]) {
        usedMarks.delete(mark);
        if (!usedMarks.size) {
          deleteCache(input);
        }
      }
    };
  };
  const store: FetchStore<Input, Result> = {
    prefetch,
    evict,
    use,
    getResult: (input: Input) => {
      const getResult = getCache(input)?.[0];
      if (!getResult) throw new Error('call prefetch in advance');
      return getResult();
    },
  };
  return store;
}
