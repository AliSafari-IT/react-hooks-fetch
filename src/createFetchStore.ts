type FetchFunc<Result, Input> = (input: Input) => Promise<Result>;

type Listener = () => void;
type Unsubscribe = () => void;

export type FetchStore<Result, Input> = {
  getResult: (input: Input) => Result;
  prefetch: (input: Input) => void;
  subscribe: (input: Input, listener: Listener) => Unsubscribe;
};

/**
 * create fetch store
 *
 * @example
 * import { createFetchStore } from 'react-hooks-fetch';
 *
 * const fetchFunc = async (userId) => (await fetch(`https://reqres.in/api/users/${userId}?delay=3`)).json();
 * const store = createFetchStore(fetchFunc);
 */
export function createFetchStore<Result, Input>(
  fetchFunc: FetchFunc<Result, Input>,
) {
  type CacheItem = {
    getResult: () => Result;
    listeners: Set<Listener>;
  };
  const cache = new Map<Input, CacheItem>();
  const createGetResult = (input: Input) => {
    let promise: Promise<void> | null = null;
    let result: Result | null = null;
    let error: Error | null = null;
    promise = (async () => {
      try {
        result = await fetchFunc(input);
      } catch (e) {
        error = e;
      } finally {
        promise = null;
      }
    })();
    const getResult = () => {
      if (promise) throw promise;
      if (error !== null) throw error;
      if (result === null) throw new Error(); // should not happen
      return result;
    };
    return getResult;
  };
  const prefetch = (input: Input) => {
    const cacheItem = cache.get(input);
    if (cacheItem) {
      cacheItem.getResult = createGetResult(input);
      cacheItem.listeners.forEach((listener) => listener());
    } else {
      cache.set(input, {
        getResult: createGetResult(input),
        listeners: new Set<Listener>(),
      });
    }
  };
  const subscribe = (input: Input, listener: () => void) => {
    const cacheItem = cache.get(input);
    if (!cacheItem) throw new Error('needs to prefetch in advance');
    const { listeners } = cacheItem;
    listeners.add(listener);
    return () => {
      listeners.delete(listener);
      if (listeners.size === 0 && cacheItem === cache.get(input)) {
        cache.delete(input);
      }
    };
  };
  const store: FetchStore<Result, Input> = {
    getResult: (input: Input) => {
      const cacheItem = cache.get(input);
      if (!cacheItem) throw new Error('needs to prefetch in advance');
      return cacheItem.getResult();
    },
    prefetch,
    subscribe,
  };
  return store;
}
