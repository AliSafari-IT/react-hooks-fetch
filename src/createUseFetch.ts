/* eslint arrow-parens: off */ // FIXME why does it complain?

import { useCallback, useState } from 'react';
import { prefetch } from 'react-suspense-fetch';

type FetchFunc<Result, Input> = (input: Input) => Promise<Result>;

export const createUseFetch = <Result extends object, Input>(
  fetchFunc: FetchFunc<Result, Input>,
  initialInput: Input,
) => {
  const initialResult = prefetch(fetchFunc, initialInput);
  const useFetch = () => {
    const [result, setResult] = useState(initialResult);
    const refetch = useCallback((input) => {
      setResult(prefetch(fetchFunc, input));
    }, []);
    return { result, refetch };
  };
  return useFetch;
};

export const createUseFetchWithoutPrefetch = <Result extends object, Input>(
  fetchFunc: FetchFunc<Result, Input>,
) => {
  const useFetch = () => {
    const [result, setResult] = useState<Result | null>(null);
    const refetch = useCallback((input) => {
      setResult(prefetch(fetchFunc, input));
    }, []);
    return { result, refetch };
  };
  return useFetch;
};
