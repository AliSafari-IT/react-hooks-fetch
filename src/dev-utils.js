/* eslint-disable no-console */

let lastInput = null;
const calls = [];
export const checkInfiniteLoop = (input) => {
  if (lastInput === input) {
    const now = Date.now();
    if (calls.length > 10) {
      if (calls[0] > now - 100) {
        console.log('Too many invocations in a short period. You probably forgot to memoize opts.');
      }
      calls.splice(0);
    } else {
      calls.push(now);
    }
  } else {
    lastInput = input;
    if (calls.length) {
      calls.splice(0);
    }
  }
};
