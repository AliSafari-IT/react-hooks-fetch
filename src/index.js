import { useCallback, useState } from 'react';

const createFetchFunc = (input) => {
  if (typeof input === 'function') return input;
  return async () => {
    const response = await fetch(input);
    const data = await response.json();
    return data;
  };
};

export const createAsync = (input) => {
  const fetchFunc = createFetchFunc(input);
  const state = { pending: true };
  state.promise = (async () => {
    try {
      state.data = await fetchFunc();
    } catch (e) {
      state.error = e;
    } finally {
      state.pending = false;
    }
  })();
  return {
    get data() {
      if (state.pending) throw state.promise;
      if (state.error) throw state.error;
      return state.data;
    },
  };
};

export const createStatic = data => ({
  get data() {
    return data;
  },
});

export const useAsync = (initialResult) => {
  const [result, setResult] = useState(initialResult);
  result.refetch = useCallback((nextInput) => {
    const nextResult = createAsync(nextInput);
    setResult(nextResult);
  }, []);
  return result;
};
