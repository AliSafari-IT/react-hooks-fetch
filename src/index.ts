export { ErrorBoundary } from './ErrorBoundary';
export { useFetch, useFetchWithoutPrefetch } from './useFetch';
export { prefetch } from 'react-suspense-fetch';
