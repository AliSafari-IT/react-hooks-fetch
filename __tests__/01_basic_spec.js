import React, { StrictMode, Suspense } from 'react';
import { render, cleanup } from 'react-testing-library';

import { useFetch } from '../src/index';

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

describe('basic spec', () => {
  afterEach(cleanup);

  it('should have a function', () => {
    expect(useFetch).toBeDefined();
  });

  it('should create a component', async () => {
    fetch.mockResponse(JSON.stringify('test data'));
    const DisplayRemoteData = () => {
      const { error, data } = useFetch('http://...');
      if (error) return <span>Error: {error.message}</span>;
      if (!data) return null; // this is important
      return <span>RemoteData: {data}</span>;
    };
    const App = () => (
      <StrictMode>
        <Suspense fallback={<span>Loading...</span>}>
          <DisplayRemoteData />
        </Suspense>
      </StrictMode>
    );
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    await sleep(10); // not ideal...
    expect(container).toMatchSnapshot();
  });
});
