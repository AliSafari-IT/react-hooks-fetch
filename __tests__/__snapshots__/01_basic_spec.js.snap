// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create a component 1`] = `
<div>
  <span>
    Loading...
  </span>
</div>
`;

exports[`basic spec should create a component 2`] = `
<div>
  <span>
    RemoteData: 
    test data
  </span>
</div>
`;
